File: flutter/lib/src/web/room_handle.dart
import 'dart:js_interop';

import '../interface/connection_handle.dart';
import '../interface/media_stream_settings.dart' as base_settings;
import '../interface/media_track.dart';
import '../interface/reconnect_handle.dart';
import '../interface/room_close_reason.dart';
import '../interface/room_handle.dart';
import '../util/move_semantic.dart';
import 'connection_handle.dart';
import 'exceptions.dart';
import 'jason_wasm.dart' as wasm;
import 'local_media_track.dart';
import 'media_stream_settings.dart';
import 'reconnect_handle.dart';
import 'room_close_reason.dart';

class WebRoomHandle implements RoomHandle {
  late wasm.RoomHandle obj;

  WebRoomHandle(this.obj);

  @override
  Future<void> join(String token) async {
    await fallibleFuture(obj.join(token).toDart);
  }

  @override
  Future<void> setLocalMediaSettings(
    base_settings.MediaStreamSettings settings,
    bool stopFirst,
    bool rollbackOnFail,
  ) async {
    await fallibleFuture(
      obj
          .set_local_media_settings(
            (settings as MediaStreamSettings).obj,
            stopFirst,
            rollbackOnFail,
          )
          .toDart,
    );
  }

  @override
  Future<void> muteAudio([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.mute_audio(kind?.index).toDart);
  }

  @override
  Future<void> unmuteAudio([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.unmute_audio(kind?.index).toDart);
  }

  @override
  Future<void> enableAudio([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.enable_audio(kind?.index).toDart);
  }

  @override
  Future<void> disableAudio([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.disable_audio(kind?.index).toDart);
  }

  @override
  Future<void> muteVideo([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.mute_video(kind?.index).toDart);
  }

  @override
  Future<void> unmuteVideo([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.unmute_video(kind?.index).toDart);
  }

  @override
  Future<void> enableVideo([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.enable_video(kind?.index).toDart);
  }

  @override
  Future<void> disableVideo([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.disable_video(kind?.index).toDart);
  }

  @override
  Future<void> enableRemoteAudio([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.enable_remote_audio(kind?.index).toDart);
  }

  @override
  Future<void> disableRemoteAudio([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.disable_remote_audio(kind?.index).toDart);
  }

  @override
  Future<void> enableRemoteVideo([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.enable_remote_video(kind?.index).toDart);
  }

  @override
  Future<void> disableRemoteVideo([MediaSourceKind? kind]) async {
    await fallibleFuture(obj.disable_remote_video(kind?.index).toDart);
  }

  @override
  void onNewConnection(void Function(ConnectionHandle) f) {
    void fn(JSAny? handle) =>
        f(WebConnectionHandle(handle as wasm.ConnectionHandle));

    fallibleFunction(() => obj.on_new_connection(fn.toJS));
  }

  @override
  void onClose(void Function(RoomCloseReason) f) {
    void fn(JSAny? reason) =>
        f(WebRoomCloseReason(reason as wasm.RoomCloseReason));
    fallibleFunction(() => obj.on_close(fn.toJS));
  }

  @override
  void onLocalTrack(void Function(LocalMediaTrack) f) {
    void fn(JSAny? track) =>
        f(WebLocalMediaTrack(track as wasm.LocalMediaTrack));
    fallibleFunction(() => obj.on_local_track(fn.toJS));
  }

  @override
  void onConnectionLoss(void Function(ReconnectHandle) f) {
    void fn(JSAny? handle) =>
        f(WebReconnectHandle(handle as wasm.ReconnectHandle));
    fallibleFunction(() => obj.on_connection_loss(fn.toJS));
  }

  @override
  void onFailedLocalMedia(void Function(Object) f) {
    void fn(JSAny? e) => f(convertException(e));
    fallibleFunction(() => obj.on_failed_local_media(fn.toJS));
  }

  @moveSemantics
  @override
  void free() {
    obj.free();
  }
}

File: src/api/dart/mock/room_handle.rs
#![allow(
    clippy::needless_pass_by_value,
    clippy::unused_async,
    clippy::unused_self,
    missing_copy_implementations
)]

use std::future::Future;

use futures::future;
use tracerr::Traced;

use crate::{
    api::{
        dart::utils::DartError, ConnectionHandle, LocalMediaTrack,
        MediaStreamSettings, ReconnectHandle,
    },
    media::MediaSourceKind,
    peer::{LocalMediaError, TracksRequestError, UpdateLocalStreamError},
    platform,
    room::{
        ChangeMediaStateError, ConstraintsUpdateError, HandleDetachedError,
        RoomCloseReason, RoomJoinError,
    },
    rpc::{ClientDisconnect, CloseReason, ConnectionInfo},
};

/// Alias for a [`Result`] related to [`MediaState`] update functions.
type ChangeMediaStateResult = Result<(), Traced<ChangeMediaStateError>>;

#[derive(Clone, Debug)]
pub struct RoomHandle(pub u8);

#[allow(clippy::missing_errors_doc, clippy::unused_async)]
impl RoomHandle {
    pub fn on_new_connection(
        &self,
        cb: platform::Function<ConnectionHandle>,
    ) -> Result<(), Traced<HandleDetachedError>> {
        cb.call1(ConnectionHandle(0));
        Ok(())
    }

    pub fn on_close(
        &self,
        cb: platform::Function<RoomCloseReason>,
    ) -> Result<(), Traced<HandleDetachedError>> {
        println!("RAZ CALL5");
        let a = CloseReason::ByClient {
            is_err: true,
            reason: ClientDisconnect::RpcClientUnexpectedlyDropped,
        };
        println!("RAZ CALL6");
        let b = RoomCloseReason::new(a);
        println!("RAZ CALL7");
        cb.call1(b);
        println!("RAZ CALL8");
        Ok(())
    }

    pub fn on_local_track(
        &self,
        cb: platform::Function<LocalMediaTrack>,
    ) -> Result<(), Traced<HandleDetachedError>> {
        cb.call1(LocalMediaTrack(0));
        Ok(())
    }

    pub fn on_connection_loss(
        &self,
        cb: platform::Function<ReconnectHandle>,
    ) -> Result<(), Traced<HandleDetachedError>> {
        cb.call1(ReconnectHandle(0));
        Ok(())
    }

    pub async fn join(
        &self,
        token: String,
    ) -> Result<(), Traced<RoomJoinError>> {
        token
            .parse::<ConnectionInfo>()
            .map_err(tracerr::map_from_and_wrap!())
            .map(drop)
    }

    pub fn on_failed_local_media(
        &self,
        cb: platform::Function<DartError>,
    ) -> Result<(), Traced<HandleDetachedError>> {
        cb.call1(
            tracerr::new!(LocalMediaError::UpdateLocalStreamError(
                UpdateLocalStreamError::InvalidLocalTracks(
                    TracksRequestError::NoTracks,
                ),
            ))
            .into(),
        );
        Ok(())
    }

    pub async fn set_local_media_settings(
        &self,
        _settings: MediaStreamSettings,
        _stop_first: bool,
        _rollback_on_fail: bool,
    ) -> Result<(), ConstraintsUpdateError> {
        Ok(())
    }

    pub fn mute_audio(
        &self,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }

    pub fn unmute_audio(
        &self,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }

    pub fn enable_audio(
        &self,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }

    pub fn disable_audio(
        &self,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }

    pub fn mute_video(
        &self,
        source_kind: Option<MediaSourceKind>,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        assert_eq!(source_kind, None);
        future::ok(())
    }

    pub fn unmute_video(
        &self,
        source_kind: Option<MediaSourceKind>,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        assert_eq!(source_kind, Some(MediaSourceKind::Display));
        future::ok(())
    }

    pub fn enable_video(
        &self,
        source_kind: Option<MediaSourceKind>,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        assert_eq!(source_kind, Some(MediaSourceKind::Device));
        future::ok(())
    }

    pub fn disable_video(
        &self,
        source_kind: Option<MediaSourceKind>,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        assert_eq!(source_kind, Some(MediaSourceKind::Display));
        future::ok(())
    }

    pub fn enable_remote_audio(
        &self,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }

    pub fn disable_remote_audio(
        &self,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }

    pub fn enable_remote_video(
        &self,
        _: Option<MediaSourceKind>,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::err(tracerr::new!(ChangeMediaStateError::Detached))
    }

    pub fn disable_remote_video(
        &self,
        _: Option<MediaSourceKind>,
    ) -> impl Future<Output = ChangeMediaStateResult> + 'static {
        future::ok(())
    }
}
